File: java/client/src/main/java/glide/api/models/commands/FlushMode.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import glide.api.GlideClient;
import glide.api.GlideClusterClient;
import glide.api.models.configuration.RequestRoutingConfiguration.Route;
import glide.api.models.configuration.RequestRoutingConfiguration.SingleNodeRoute;

// TODO add links to script flush
/**
 * Defines flushing mode for:
 *
 * <ul>
 *   <li><code>FLUSHALL</code> command implemented by {@link GlideClient#flushall(FlushMode)},
 *       {@link GlideClusterClient#flushall(FlushMode)}, and {@link
 *       GlideClusterClient#flushall(FlushMode, SingleNodeRoute)}.
 *   <li><code>FLUSHDB</code> command implemented by {@link GlideClient#flushdb(FlushMode)}, {@link
 *       GlideClusterClient#flushdb(FlushMode)}, and {@link GlideClusterClient#flushdb(FlushMode,
 *       SingleNodeRoute)}.
 *   <li><code>FUNCTION FLUSH</code> command implemented by {@link
 *       GlideClient#functionFlush(FlushMode)}, {@link GlideClusterClient#functionFlush(FlushMode)},
 *       and {@link GlideClusterClient#functionFlush(FlushMode, Route)}.
 *   <li><code>SCRIPT FLUSH</code> command implemented by {@link GlideClient#scriptFlush(FlushMode)}
 *       and {@link GlideClusterClient#scriptFlush(FlushMode, Route)}
 * </ul>
 *
 * @see <a href="https://valkey.io/commands/flushall/">flushall</a>, <a
 *     href="https://valkey.io/commands/flushdb/">flushdb</a>, and <a
 *     href="https://valkey.io/commands/function-flush/">function flush</a> at valkey.io
 */
public enum FlushMode {
    /**
     * Flushes synchronously.
     *
     * @since Valkey 6.2 and above.
     */
    SYNC,
    /** Flushes asynchronously. */
    ASYNC;

    // Pre-allocated argument arrays to reduce allocations
    private static final String[] SYNC_ARGS = new String[] {"SYNC"};
    private static final String[] ASYNC_ARGS = new String[] {"ASYNC"};

    /**
     * Returns the command arguments for this flush mode. Uses pre-allocated arrays to reduce GC
     * pressure.
     *
     * @return Array containing the flush mode as a string argument
     */
    public String[] toArgs() {
        return this == SYNC ? SYNC_ARGS : ASYNC_ARGS;
    }
}

File: node/src/commands/FlushMode.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

// Import below added to fix up the TSdoc link, but eslint blames for unused import.
/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
import { GlideClient } from "src/GlideClient";

/**
 * Defines flushing mode for {@link GlideClient.flushall|flushall} and {@link GlideClient.flushdb|flushdb} commands.
 *
 * See https://valkey.io/commands/flushall/ and https://valkey.io/commands/flushdb/ for details.
 */
export enum FlushMode {
    /**
     * Flushes synchronously.
     *
     * since Valkey version 6.2.0.
     */
    SYNC = "SYNC",
    /** Flushes asynchronously. */
    ASYNC = "ASYNC",
}
